### ISSUE TITLE
Dynamically load .class files

### ISSUE THREAD

**Issue by Cyrik** [nrepl/nrepl#262]
**Is your feature request related to a problem? Please describe.**

When .class files in the classpath get updated, they don't get reloaded.

**Describe the solution you'd like**

When a .class file on the path changes, it should be reloaded. This should probably only work for files outside jars, just to reduce performance impact.
The only way to get this behavior is a new classLoader, similar to Clojure's own dynamicClassLoader.

**Describe alternatives you've considered**

The other way to handle this case is to restart nrepl or to build your own classLoader, both of which seem like bad solutions. The problem could also be ignored, since most people won't be compiling class files on the fly.

**Additional context**

Based on this Twitter [thread](https://twitter.com/pappapez/status/1476673398914830338)

**Comment by bbatsov** [comment]
I assume that's a problem in projects with mixed Clojure & Java sources, right?

I'll defer to @plexus about whether we can tackle this in a practical manner, given he's our resident classloader expert and has been working on simplifying classloader-related logic in nREPL.

**Comment by Cyrik** [comment]
Yes, this is only a problem if you mix sources and recompile the java part.

**Comment by plexus** [comment]
There is actually a project that solves this exact problem: [Virgil](https://github.com/ztellman/virgil), it watches java files, and re-compiles and hot-reloads them when they change. Virgil then also reloads the Clojure sources that rely on that class, which shows the potential pitfalls, existing instances of the class won't show the new behavior. I'm not sure what happens if a Java class that references the reloaded class instantiates a new instance after the reload, I think it's quite possible it might still get an instance of the old class.

As mentioned you can do this with DynamicClassloader, that's why it exists, no need for a custom classloader.

In any case I don't really see how this is an nREPL concern. This can be handled orthogonally by something like Virgil, and they've already put quite some effort into handling all the details, so probably better to point people there, and possibly adopt it into clj-commons if improvements are needed and maintenance is an issue.

**Comment by Cyrik** [comment]
Thank you @plexus, I'll have a look at Virgil.

**Comment by bbatsov** [comment]
I guess we can just mention Virgil somewhere in the docs and call it a day.

**Comment by Cyrik** [comment]
Yes, @bbatsov. We'll add a section about it in the calva docs, but it probably wouldn't hurt for nrepl to also mention it somewhere.

**Comment by bbatsov** [comment]
Under `usage/misc` or in the FAQ.
